README.md
![alt text](https://github.com/rpidanny/hima.js/raw/master/banner.png "hima.js")

![gh-actions](https://github.com/rpidanny/hima.js/workflows/Publish/badge.svg) [![codecov](https://codecov.io/gh/rpidanny/hima.js/branch/master/graph/badge.svg)](https://codecov.io/gh/rpidanny/hima.js)

## Highlights

- Promise API
- CLI
- TypeScript Support
- Actively Maintained
- Download Single Image
- Download Multiple Images between two dates
- Create Timelapse Video

## Install

```bash
$ brew install graphicsmagick
$ brew install ffmpeg
$ npm install --save @rpidanny/hima
```

`ffmpeg` is required to create the timelapse video. Feel free to skip this if you don't want to create timelapse videos.

## Usage

### CLI

Install package globally using:

`$ npm i -g @rpidanny/hima`

```bash
Usage: hima image [options]

Download single image

Options:
  --out <path>    Output file (default: "./<currentdate>.jpg")
  --date <date>   Date in yyyy/mm/dd hh:mm:ssZ (default: "latest")
  --zoom <level>  Zoom level. 1-3 for IR. 1-5 for color (default: "1")
  --ir            Download Infrared Image (default: false)
  --batch-size    How many tiles to download in parallel?
  --debug         Enable debug logs? (default: false)
  --quiet         Disable all logs? (default: false)
  -h, --help      display help for command
```

```bash
Usage: hima images [options]

Download multiple images

Options:
  --out <path>          Output directory (default: "./")
  --start-date <date>   Date in yyyy/mm/dd hh:mm:ssZ
  --end-date <date>     Date in yyyy/mm/dd hh:mm:ssZ
  --interval <minutes>  Interval between two images (default: "10")
  --zoom <level>        Zoom level. 1-3 for IR. 1-5 for color (default: "1")
  --ir                  Download Infrared Image (default: false)
  --batch-size          How many tiles to download in parallel?
  --debug               Enable debug logs? (default: false)
  --quiet               Disable all logs? (default: false)
  -h, --help            display help for command
```

```bash
Usage: hima timelapse [options]

Create timelapse video

Options:
  --out <path>            Output file (default: "./<currentdate>.mp4")
  --start-date <date>     Date in yyyy/mm/dd hh:mm:ssZ
  --end-date <date>       Date in yyyy/mm/dd hh:mm:ssZ
  --interval <minutes>    Interval between two images (default: "10")
  --quality <resolution>  Resolution. 480, 720, 1080, 1440, 2160 (default: "1080")
  --fps <rate>            Framerate of the video (default: "25")
  --ir                    Download Infrared Image (default: false)
  --batch-size            How many tiles to download in parallel?
  --debug                 Enable debug logs? (default: false)
  --quiet                 Disable all logs? (default: false)
  -h, --help              display help for command
```

### API

#### downloadImage(options?)

##### options

Type: `object`

| key          | default        | type              | description       |
| ------------ | -------------- | ----------------- | ----------------- |
| `date`       | `latest`       | `string` / `date` | String in `yyyy/mm/dd hh:mm:ssZ` or a JS `Date` object.       |
| `zoom`       | `1`            | `number`          | zoom level. 1-3 for IR and 1-5 for color |
| `infrared`   | `false`        | `boolean`         | color image or IR image? |
| `output`     | `./currentdate.jpg`           | `string`          | Output file.      |
| `batchSize`  | `20`           | `number`          | How many tiles to download in parallel? If you get `ECONNRESET`, try lowering the `batchSize`. |
| `debug`      | `false`        | `boolean`         | enable logs?      |
| `timeout`    | `{ connect: 15000, response: 15000, request: 30000 }` | `object`    | [got timeout](https://github.com/sindresorhus/got#timeout)   |
| `progress`   | `None`         | `function`        | A callback function that is called on progress update. Receives two parameters: (`completed`, `total`) |

#### Example

```js
import { downloadImage } from '@rpidanny/hima'

downloadImage({
  zoom: 1,
  batchSize: 20,
  date: '2019/10/21 18:30:20Z', // 2020/02/14
  debug: true,
  progress: (completed, total) => console.log(`Completed ${completed}/${total}`)
})
  .then(console.log)
  .catch(console.error)

```

#### downloadImages(options?)

##### options

Type: `object`

| key          | default  | type              | description       | required |
| ------------ | -------- | ----------------- | ----------------- | -------- |
| `startDate`  | none     | `string` / `date` | String in `yyyy/mm/dd hh:mm:ssZ` or a JS `Date` object. | `yes` |
| `endDate`    | none     | `string` / `date` | String in `yyyy/mm/dd hh:mm:ssZ` or a JS `Date` object. | `yes` |
| `interval`   | `10`     | `number`          | Interval between two images in minutes. (min: 10) | `no` |
| `zoom`       | `1`      | `number`          | zoom level. 1-3 for IR and 1-5 for color | `no` |
| `infrared`   | `false`  | `boolean`         | color image or IR image? | `no` |
| `output`     | `./`     | `string`          | Output file.      | `no` |
| `batchSize`  | `20`     | `number`          | How many tiles to download in parallel? If you get `ECONNRESET`, try lowering the `batchSize`. | `no` |
| `debug`      | `false`  | `boolean`         | enable logs?      | `no` |
| `timeout`    | `{ connect: 15000, response: 15000, request: 30000 }` | `object`    | [got timeout](https://github.com/sindresorhus/got#timeout)   | `no` |
| `progress`   | `None`   | `function`        | A callback function that is called on progress update. Receives two parameters: (`completed`, `total`) | `no` |

#### Example

```js
import { downloadImages, intervals } from '@rpidanny/hima'

downloadImages({
  zoom: 2,
  batchSize: 5,
  startDate: '2020/02/14 05:00:00Z',
  endDate: '2020/03/14 05:00:00Z',
  interval: intervals.DAY, // 1440 minutes
  debug: true,
  output: './images',
  progress: (completed, total) => console.log(`Completed ${completed}/${total}`)
})
  .then(console.log)
  .catch(console.error)

```

#### createTimelapse(options?)

##### options

Type: `object`

| key          | default        | type              | description       | required |
| ------------ | -------------- | ----------------- | ----------------- | -------- |
| `startDate`  | none           | `string` / `date` | String in `yyyy/mm/dd hh:mm:ssZ` or a JS `Date` object.       | `yes` |
| `endDate`    | none           | `string` / `date` | String in `yyyy/mm/dd hh:mm:ssZ` or a JS `Date` object.       | `yes` |
| `interval`   | `10`           | `number`          | Interval between two images in minutes. (min: 10) | `no` |
| `quality`    | `1080`         | `string`          | Resolution. 480, 720, 1080, 1440, 2160 | `no` |
| `fps`        | `25`           | `number`          | Framerate of the video. | `no` |
| `infrared`   | `false`        | `boolean`         | color image or IR image? | `no` |
| `output`     | `./<currentdate>.mp4` | `string`   | Output file.      | `no` |
| `batchSize`  | `20`           | `number`          | How many tiles to download in parallel? If you get `ECONNRESET`, try lowering the `batchSize`. | `no` |
| `debug`      | `false`        | `boolean`         | enable logs?      | `no` |
| `timeout`    | `{ connect: 15000, response: 15000, request: 30000 }` | `object`    | [got timeout](https://github.com/sindresorhus/got#timeout)   | `no` |
| `progress`   | `None`         | `function`        | A callback function that is called on progress update. Receives two parameters: (`completed`, `total`) | `no` |

#### Example

```js
import { createTimelapse, intervals } from '@rpidanny/hima'

createTimelapse({
  quality: '2160',
  batchSize: 64,
  startDate: '2019/10/14 05:00:00Z',
  endDate: '2020/03/14 05:00:00Z',
  interval: intervals.DAY, // 1440 minutes
  debug: true,
  output: './video.mp4',
  progress: (c, t) => console.log(`${(c / t) * 100}% complete`),
})
  .then(console.log)
  .catch(console.error)
```

## Development

Checkout the repository locally, then:

```bash
$ npm i
$ npm run hima
```

## Contribute

Help make this package better and future proof.

- Clone the code
- Install dependencies with `npm i`
- Create a new branch with `git checkout -b new_feature`
- Commit your changes and push your branch
- Submit a PR to `master`


CHANGELOG.md
## [1.5.1](https://github.com/rpidanny/hima.js/compare/v1.5.0...v1.5.1) (2020-10-30)


### Bug Fixes

* github-script version [skip ci] ([#101](https://github.com/rpidanny/hima.js/issues/101)) ([602f971](https://github.com/rpidanny/hima.js/commit/602f971554533259aaa2374109fda2e276d1d482))

# [1.5.0](https://github.com/rpidanny/hima.js/compare/v1.4.0...v1.5.0) (2020-10-26)


### Features

* add agentkeepalive ([#88](https://github.com/rpidanny/hima.js/issues/88)) ([64f0a25](https://github.com/rpidanny/hima.js/commit/64f0a250bc598f681cdd641184f5629c7fcc9956))

# [1.4.0](https://github.com/rpidanny/hima.js/compare/v1.3.3...v1.4.0) (2020-10-26)


### Features

* add nocks on integration test ([#86](https://github.com/rpidanny/hima.js/issues/86)) ([e2a9f4a](https://github.com/rpidanny/hima.js/commit/e2a9f4a856f99fe090a36f200a14ed348e17c66a))

## [1.3.3](https://github.com/rpidanny/hima.js/compare/v1.3.2...v1.3.3) (2020-09-13)


### Bug Fixes

* baseUrl to https ([#67](https://github.com/rpidanny/hima.js/issues/67)) ([967ce5a](https://github.com/rpidanny/hima.js/commit/967ce5a0741d5fcce6158dec01da32e3decec1f4))

## [1.3.2](https://github.com/rpidanny/hima.js/compare/v1.3.1...v1.3.2) (2020-08-25)


### Bug Fixes

* github token on auto-merge ([#47](https://github.com/rpidanny/hima.js/issues/47)) ([3311c63](https://github.com/rpidanny/hima.js/commit/3311c636f741d82c48ff57ca0a8042d245ab178f))

## [1.3.1](https://github.com/rpidanny/hima.js/compare/v1.3.0...v1.3.1) (2020-08-09)


### Performance Improvements

* implement images level batching to increase performance ([#25](https://github.com/rpidanny/hima.js/issues/25)) ([8d2e490](https://github.com/rpidanny/hima.js/commit/8d2e490c780af54bbcdc0826dfed3de758ec702c))

# [1.3.0](https://github.com/rpidanny/hima.js/compare/v1.2.3...v1.3.0) (2020-08-09)


### Features

* add interval macros ([#24](https://github.com/rpidanny/hima.js/issues/24)) ([d017f3b](https://github.com/rpidanny/hima.js/commit/d017f3be93f5d573c76492d514fa25e0fa3ee2ac))

## [1.2.3](https://github.com/rpidanny/hima.js/compare/v1.2.2...v1.2.3) (2020-08-09)


### Bug Fixes

* tsc rootDirs and package.json main ([#22](https://github.com/rpidanny/hima.js/issues/22)) ([7daa659](https://github.com/rpidanny/hima.js/commit/7daa659e45a6392b8f99f2737a6054c648169c87))

## [1.2.2](https://github.com/rpidanny/hima.js/compare/v1.2.1...v1.2.2) (2020-08-09)


### Bug Fixes

* tsc output dir structure ([#21](https://github.com/rpidanny/hima.js/issues/21)) ([31c7cf4](https://github.com/rpidanny/hima.js/commit/31c7cf49c55ef191a694ab2667792b5a76d3d99c))

## [1.2.1](https://github.com/rpidanny/hima.js/compare/v1.2.0...v1.2.1) (2020-08-09)


### Bug Fixes

* update package.json main/types ([#20](https://github.com/rpidanny/hima.js/issues/20)) ([a2f720b](https://github.com/rpidanny/hima.js/commit/a2f720bd6681cae5676f9aee71efefd79c539d0b))

# [1.2.0](https://github.com/rpidanny/hima.js/compare/v1.1.0...v1.2.0) (2020-08-08)


### Features

* add timelapse functionality ([#16](https://github.com/rpidanny/hima.js/issues/16)) ([76943dc](https://github.com/rpidanny/hima.js/commit/76943dc4a7e63b9273ac72e83123a1faa055cb53)), closes [#11](https://github.com/rpidanny/hima.js/issues/11) [#13](https://github.com/rpidanny/hima.js/issues/13) [#14](https://github.com/rpidanny/hima.js/issues/14) [#15](https://github.com/rpidanny/hima.js/issues/15)

# [1.1.0](https://github.com/rpidanny/hima.js/compare/v1.0.0...v1.1.0) (2020-08-07)


### Features

* download images from range of dates ([#10](https://github.com/rpidanny/hima.js/issues/10)) ([1b46d48](https://github.com/rpidanny/hima.js/commit/1b46d4830ead6103180144cfea3ddaa688dd8499)), closes [#12](https://github.com/rpidanny/hima.js/issues/12)


src/cli.ts
import path from 'path'
import { program } from 'commander'
import cliProgress from 'cli-progress'
import colors from 'colors' // TODO: replace with chalk

import { downloadImage, downloadImages, createTimelapse } from './index'
import { ProgressFunction } from './types'

interface ImageInputOpts {
  out: string
  date: string
  zoom: string
  ir: boolean
  batchSize: string
  debug: boolean
  quiet: boolean
}

interface ImagesInputOpts {
  out: string
  startDate: string
  endDate: string
  interval: number
  zoom: string
  ir: boolean
  batchSize: string
  debug: boolean
  quiet: boolean
}

interface TimelapseInputOpts {
  out: string
  startDate: string
  endDate: string
  interval: number
  quality: string
  fps: string
  ir: boolean
  batchSize: string
  debug: boolean
  quiet: boolean
}

// TODO: Maybe disable progressbar when debug mode is enabled?
const progressBar = new cliProgress.SingleBar(
  {
    format: `{status} |${colors.cyan('{bar}')}| {percentage}% | {completed}/{totalTiles} Chunks`,
    barCompleteChar: '\u2588',
    barIncompleteChar: '\u2591',
    hideCursor: true,
  },
  cliProgress.Presets.shades_classic,
)

const errorHandler = (err: Error): void => {
  progressBar.stop()
  console.log(err)
}

const getProgressHandler = (title: string, quiet: boolean): ProgressFunction | undefined => {
  if (quiet) {
    return
  }
  progressBar.start(100, 0, {
    percentage: 0,
    status: colors.inverse(title),
    completed: 0,
    totalTiles: 0,
  })

  return (completed, totalTiles) => {
    if (completed >= totalTiles) {
      progressBar.update((completed / totalTiles) * 100, {
        status: colors.inverse.green('Download Complete'),
        completed,
        totalTiles,
      })
      return progressBar.stop()
    }
    progressBar.update((completed / totalTiles) * 100, {
      completed,
      totalTiles,
    })
  }
}

const handleImageCommand = async (inputOpts: ImageInputOpts): Promise<void> => {
  try {
    const { output } = await downloadImage({
      date: inputOpts.date,
      output: path.resolve(inputOpts.out),
      zoom: parseInt(inputOpts.zoom),
      batchSize: (inputOpts.batchSize && parseInt(inputOpts.batchSize)) || 20,
      infrared: inputOpts.ir,
      debug: inputOpts.quiet || inputOpts.debug,
      progress: getProgressHandler('Downloading Image', inputOpts.quiet),
    })

    if (!inputOpts.quiet && output) {
      console.log(colors.green.inverse(`File saved at ${output}`))
    }
  } catch (err) {
    errorHandler(err)
  }
}

const handleImagesCommand = async (inputOpts: ImagesInputOpts): Promise<void> => {
  console.log(inputOpts)
  try {
    const { images } = await downloadImages({
      startDate: inputOpts.startDate,
      endDate: inputOpts.endDate,
      interval: inputOpts.interval,
      output: path.resolve(inputOpts.out),
      zoom: parseInt(inputOpts.zoom),
      batchSize: (inputOpts.batchSize && parseInt(inputOpts.batchSize)) || 20,
      infrared: inputOpts.ir,
      debug: inputOpts.quiet || inputOpts.debug,
      progress: getProgressHandler('Downloading Images', inputOpts.quiet),
    })
    if (!inputOpts.quiet && images) {
      console.log(colors.green.inverse(`${images.length} Images saved to ${inputOpts.out}`))
    }
  } catch (err) {
    errorHandler(err)
  }
}

const handleTimelapseCommand = async (inputOpts: TimelapseInputOpts): Promise<void> => {
  try {
    const { output } = await createTimelapse({
      startDate: inputOpts.startDate,
      endDate: inputOpts.endDate,
      interval: inputOpts.interval,
      output: path.resolve(inputOpts.out),
      quality: inputOpts.quality,
      fps: parseInt(inputOpts.fps),
      batchSize: (inputOpts.batchSize && parseInt(inputOpts.batchSize)) || 20,
      infrared: inputOpts.ir,
      debug: inputOpts.quiet || inputOpts.debug,
      progress: getProgressHandler('Downloading Images', inputOpts.quiet),
    })
    if (!inputOpts.quiet && output) {
      console.log(colors.green.inverse(`Timelapse video saved to ${output}`))
    }
  } catch (err) {
    errorHandler(err)
  }
}

async function main(): Promise<void> {
  const rootPath = process.cwd()

  const defaultFile = path.resolve(rootPath, `${new Date().toISOString()}.jpg`)
  const defaultVideo = path.resolve(rootPath, `${new Date().toISOString()}.mp4`)

  program
    .command('image [command]')
    .description('Download single image')
    .option('--out <path>', 'Output file', defaultFile)
    .option('--date <date>', 'Date in yyyy/mm/dd hh:mm:ssZ', 'latest')
    .option('--zoom <level>', 'Zoom level. 1-3 for IR. 1-5 for color', '1')
    .option('--ir', 'Download Infrared Image', false)
    .option('--batch-size <number>', 'How many tiles to download in parallel?', '20')
    .option('--debug', 'Enable debug logs?', false)
    .option('--quiet', 'Disable all logs?', false)
    .action(async (_env, options) => await handleImageCommand(options))

  program
    .command('images [command]')
    .description('Download multiple image')
    .option('--out <path>', 'Output directory', rootPath)
    .option('--start-date <date>', 'Date in yyyy/mm/dd hh:mm:ssZ')
    .option('--end-date <date>', 'Date in yyyy/mm/dd hh:mm:ssZ')
    .option('--interval <minutes>', 'Interval between two images', '10')
    .option('--zoom <level>', 'Zoom level. 1-3 for IR. 1-5 for color', '1')
    .option('--ir', 'Download Infrared Image', false)
    .option('--batch-size <number>', 'How many tiles to download in parallel?', '20')
    .option('--debug', 'Enable debug logs?', false)
    .option('--quiet', 'Disable all logs?', false)
    .action(async (_env, options) => await handleImagesCommand(options))

  program
    .command('timelapse [command]')
    .description('Create timelapse video')
    .option('--out <path>', 'Output file', defaultVideo)
    .option('--start-date <date>', 'Date in yyyy/mm/dd hh:mm:ssZ')
    .option('--end-date <date>', 'Date in yyyy/mm/dd hh:mm:ssZ')
    .option('--interval <minutes>', 'Interval between two images', '10')
    .option('--quality <resolution>', 'Resolution. 480, 720, 1080, 1440, 2160', '1080')
    .option('--fps <rate>', 'Framerate of the video', '25')
    .option('--ir', 'Download Infrared Image', false)
    .option('--batch-size <number>', 'How many tiles to download in parallel?', '20')
    .option('--debug', 'Enable debug logs?', false)
    .option('--quiet', 'Disable all logs?', false)
    .action(async (_env, options) => await handleTimelapseCommand(options))

  program.parse(process.argv)
}

try {
  main()
} catch (err) {
  console.error(err)
}
